lib/msf/core/post/windows/services.rb

```diff
@@ -655,6 +663,94 @@ def meterpreter_service_list
 
           services
         end
+
+        #
+        # Start a service using sc.exe.
+        #
+        # @param name [String] Service name (not display name)
+        # @param server [String,nil] A hostname or IP address. Default is the
+        #   remote localhost.
+        #
+        # @return [Integer] 0 if service started successfully, 1 if it failed
+        #   because the service is already running, 2 if it is disabled
+        #
+        # @raise [RuntimeError] starting service failed
+        #
+        def _shell_service_start(service_name, server = nil)
+          host = server ? "\\\\#{server}" : nil
+          timeout = 62 # sc.exe default RPC connection timeout is 60 seconds
+
+          fingerprint = Rex::Text.rand_text_alphanumeric(6..8)
+
+          res = cmd_exec("sc #{host} start #{service_name} && echo #{fingerprint}", nil, timeout).to_s
+
+          raise "Could not start service #{service_name}. sc.exe returned no output." if res.blank?
+
+          code = res.split(/\r?\n/).first.scan(/ (\d+):/).flatten.first
+
+          return Error::SUCCESS if res.include?(fingerprint) && code.nil?
+
+          raise "Could not start service #{service_name.inspect}. sc.exe returned unexpected output." if code.nil?
+
+          case code.to_i
+          when Error::SERVICE_ALREADY_RUNNING
+            return 1
+          when Error::SERVICE_DISABLED
+            return 2
+          when Error::SERVICE_DOES_NOT_EXIST
+            raise "[SC] StartService: The specified service #{service_name.inspect} does not exist as an installed service."
+          when Error::RPC_S_SERVER_UNAVAILABLE
+            raise "[SC] StartService: Could not connect to RPC server #{server}"
+          else
+            status = ::WindowsError::Win32.find_by_retval(code.to_i).first
+            raise "[SC] StartService: Unhandled error: #{status.name}: #{status.description}"
+          end
+        end
+
+        #
+        # Stop a service using sc.exe.
+        #
+        # @param name [String] Service name (not display name)
+        # @param server [String,nil] A hostname or IP address. Default is the
+        #   remote localhost.
+        #
+        # @return [Integer] 0 if service stopped successfully, 1 if it failed
+        #   because the service is already stopped or disabled, 2 if it
+        #   cannot be stopped for some other reason.
+        #
+        # @raise [RuntimeError] stopping service failed
+        #
+        def _shell_service_stop(service_name, server = nil)
+          host = server ? "\\\\#{server}" : nil
+          timeout = 62 # sc.exe default RPC connection timeout is 60 seconds
+
+          fingerprint = Rex::Text.rand_text_alphanumeric(6..8)
+
+          res = cmd_exec("sc #{host} stop #{service_name} && echo #{fingerprint}", nil, timeout).to_s
```

Comment:
do we need the `to_s` here?

Reply:
Probably not as we immediately check for `nil` using `blank?` - however, we also perform multiple operations on `res` assuming that it is a string (`split` and `include?`).

Reply:
changed